### ISSUE TITLE
Exception and LuaJIT

### ISSUE THREAD

**Issue by otristan** [ThePhD/sol2#955]
Hi,

I'm having trouble understanding some stuff with LuaJIT

Sol is setup as this:

#ifdef USE_LUA_JIT
#define SOL_USING_CXX_LUAJIT 1
#else
#define SOL_USE_BOOST 1
#define SOL_USING_CXX_LUA 1
#endif

#define SOL_SAFE_NUMERICS  1
#define SOL_CHECK_ARGUMENTS 1
#define SOL_EXCEPTIONS_SAFE_PROPAGATION 1

Using regular Lua, I'm using unsafe_function, script_file and catching exception myself.
So far this has worked perfectly.
When using LuaJIT (trunk git), in the case of a typo in the script file, I get some exception in script_file that I can only catch with catch(...)
(OSX x64 so full interoperability with exception)

Tried that:
```
template <typename F, typename... Args>
    void exception_wrapper(const F& callback, Args&&... args)
    {
      try
      {
        callback(std::forward<Args>(args)...);
      }
      catch (const sol::error& e)
      {
        spdlog::info("sol::error: {}", e.what());
      }
      catch (const std::runtime_error& e)
      {
        spdlog::info("std::runtime error: {}", e.what());
      }
      catch (const std::exception& e)
      {
        spdlog::info("std::exception error: {}", e.what());
      }
      catch (const std::string &s)
      {
        spdlog::info("std::string error: {}", s);
      }
      catch (const char* cs)
      {
        spdlog::info("char* error: {}", cs);
      }
      catch (...)
      {
        try {
          auto eptr = std::current_exception();
          if (eptr)
            std::rethrow_exception(eptr);
          else
            spdlog::info("unknown exception");
        } catch (std::exception& e)
        {
          spdlog::info("std::exception error: {}", e.what());
        }
        
      }
    }
```

If I turn on SOL_SAFE_FUNCTION, I now get an exception through script_file (script_throw_on_error) and am able to catch it, although I was understanding that the point of safe_function was that exception was catched internally.

Now if I have an error in a function that I call through c++, I see an error in debug, thanks to sol own print in default_at_panic but then I mostly get a crash using either safe or unsafe function afterwards

It's seem similar to this issue
https://github.com/ThePhD/sol2/issues/632

So basically, I am wondering what is the best to do.
setup my own default_at_panic and use or not SOL_SAFE_FUNCTION or something else

I'm puzzled.

Thanks !

**Comment by ThePhD** [comment]
`unsafe_function` is exactly that: unsafe. It assumes success. If success doesn't happen, it blows you up.

Letting an exception escape the interface is dangerous. Exceptions escape Lua mean it can leave the VM is an unpredictable state. Also, did you actually build your LuaJIT from scratch as C++? That's what the `SOL_USING_CXX_LUA` flag is for.

I would recommend using `sol::protected_function` and then reading from the error after doing a call that way. There are examples in the documentation and the `examples` folder as well.

**Comment by otristan** [comment]
Yep that what I have ended up doing which works
- only use safe function
- try catch around script_file
- my own default_at_panic

I've my build my own luaJIT but with regular make which uses clang so no idea regarding
SOL_USING_CXX_LUAJIT.
It wouldn't work otherwise no ?
